### ISSUE TITLE
Verbose mode for ltfs_ordered_copy

### ISSUE THREAD

**Issue by svein** [LinearTapeFileSystem/ltfs#200]
The only thing holding me back with the aging ltfscp is the lack of verbose mode for ltfs_ordered_copy. A --log option would be great also.

**Comment by piste-jp-ibm** [comment]
As you see the code, I would like to have this kind of functionality. But I don't think `ltfscp`'s implementation is teh best of it.

What do you think which information is useful for users? 

https://github.com/LinearTapeFileSystem/ltfs/blob/623f476aaecf0fa55f247927b5f5754acb983ab8/src/utils/ltfs_ordered_copy#L78

**Comment by svein** [comment]
It might not be the best no, it could be a lot cleaner. Omitting some of the command codes perhaps (ILT30505I Copy)?
The thing I like from ltfscp output is:
* Full path (from/path/file destination/path/file)
* The summary at the end with number of files copied and the return code, though the last part I could get from the shell obviously.

But obviously there is a lot of other ways to present the data which would be fine. like cp -v does it or rsync
cp doesn't have the summary though, and I don't think rsync shows the destination path.

**Comment by piste-jp-ibm** [comment]
`ltfs_ordered_copy` provides a similar functionality of `ltfscp`, but it is not retreated into `ltfscp` at all. Implementation is completely designed from clean sheet.

As a basic strategy, I don't want to create a clone of `ltfscp`. But I will add something good things in `ltfscp`.

I would like to add following things as you suggested. What do you think?

* Print progress information to `stderr` when no option is specified. Like `Copying files (10/1000)`
* Print source destination to `stderr` when `-v` option is specified. Like `Copying file /foo-s/bar-s -> /foo-d/bar-d`
* Return 0 when all file copy is successfully done
* Return 1 when one or more file copy failes
* Return 2 for other various errors

I believe printing a result is not necessary if you can get a designed return code by `$?` in shell script.

**Comment by svein** [comment]
Printing to stderr is probably what I liked least of `ltfscp`. I would prefer normal information prints to `stdout`.
Perhaps have a `-q` option also, which is completely quiet and only prints errors to `stderr`?

As for return codes I completely agree with you on that.

**Comment by piste-jp-ibm** [comment]
@svein 

I have commited a quick prototype on https://github.com/piste-jp-ibm/ltfs/tree/oc_enhance. Could you try it if you have a chance?

The prototype supports
  * Set verbose level with `--verbose` (default is 4)
  * Verbosity option with `-v`(set verbose level 5 and this prints file name messages at each copy)
  * Quiet option with `-q` (set verbose level 0 and print nothing)
  * Correct return code (on 0 success all files, on 1 fail 1 ore more files, on 2 usage error)

The prototype does not support
  * Copy progress message on `-v` (verbose level 4 or larger)

**Comment by piste-jp-ibm** [comment]
I have committed progress message support on https://github.com/piste-jp-ibm/ltfs/tree/oc_enhance. 

I will make a PR once my UT is finished. I can accept comments anytime in this thread.

**Comment by svein** [comment]
Hi,
I have tested a bit with different values and it seems to be working. Though what is the numbers between source and destination? Is it inode?

`--verbose` need a input or else it throws an error, but that's not what the help says:
`ltfs_ordered_copy: error: argument --verbose: expected one argument`

It's a bit confusing that `-v `don't take a value and `--verbose` does, also that the default value is different. But that's nitpicking.

For me the the value of the output is what I can do with it later. For instance to verify that the files is in the correct place, therefore the output needs to be easy to parse. I would like to be able to pull out the source path or the destination path, so perhaps have the ability to have output like:
"/my/cr a cy¥£/path.log" -> "/some destination/my/cr a cy¥£/path.log"
would be really helpful. Since the tool only copy

I'll test more when I'm back at work.

Thank you for your great work on this.

**Comment by piste-jp-ibm** [comment]
```
usage: ltfs_ordered_copy [-h] [-p] [-r] [-t TARGET_DIRECTORY]
....
                        only stdin source
  -a, --all             achieve files recursively and preserve attributes
  -v                    Verbose output. Set VERBOSE level 5
  --verbose VERBOSE     Configure verbosity of logger. VERBOSE shall be 0-6.
                        default is 4
  -q, --quiet           No message output
```

My intention is 

* `-v` and `--verbose` is not a same option
* `-v` is the option to set the verbose level 5 automatically
* `--verbose` is the option to set to set the verbose level to specified level

Because I believe almost all users want to use just `-v` to use little bit higher verbosity for their convenience. `-v` is as same as `--verbose 5`. And neither `-v` nor `--verbose` is not specified, verbose level is 4.

**Comment by piste-jp-ibm** [comment]
I have updated the copy message like `"/my/cr a cy¥£/path.log" -> "/some destination/my/cr a cy¥£/path.log"`.

**Comment by piste-jp-ibm** [comment]
_Though what is the numbers between source and destination? Is it inode?_

Sorry, I can't get you. What **numbers** are you saying?

**Comment by svein** [comment]
The numbers in brackets.
```
 ./ltfs_ordered_copy -v -r 
Copying "/mnt/ltfs/A00236JD/testfolder/test2/deleteme.txt" (12326) -> "/gpfs/gpfs0/ltfstest/deleteme.txt"
Copying "/mnt/ltfs/A00236JD/testfolder/test2/1/FolderStructureReport_2020-06-04.txt" (12327) -> "/gpfs/gpfs0/ltfstest/1/FolderStructureReport_2020-06-04.txt"
Copying "/mnt/ltfs/A00236JD/testfolder/test2/2/test.docx" (29988) -> "/gpfs/gpfs0/ltfstest/2/test.docx"
Copying "/mnt/ltfs/A00236JD/testfolder/test2/3/summary.txt" (290883) -> "/gpfs/gpfs0/ltfstest/3/summary.txt"
```

**Comment by svein** [comment]
I was going to create a new ticket with a bug I found, but it seems to be in this branch you introduced it.

It seems that `ltfs_ordered_copy` crashes when it encounters attributes it can't handle. For instance 0byte files. The `master `branch throws a warning for these files, but continues. The `oc_enhance` crashes.

 ```
./ltfs_ordered_copy -r /mnt/ltfs/A00236JD /gpfs/gpfs0/ltfstest/
Tape order aware copy for LTFS
Destination /gpfs/gpfs0/ltfstest/ is not LTFS
Failed to get attribute of "/mnt/ltfs/A00236JD/1/2/3/ref04.log": [Errno 61] No data available
Failed to get attribute of "/mnt/ltfs/A00236JD/1/2/3/ref05.log": [Errno 61] No data available
Traceback (most recent call last):
  File "./ltfs_ordered_copy", line 398, in <module>
    copyq.walk_dir(s, dst, args.p, args.all)
  File "./ltfs_ordered_copy", line 172, in walk_dir
    self.add_copy_item(c)
  File "./ltfs_ordered_copy", line 142, in add_copy_item
    if tail.size >= HalfMegaByte:
NameError: global name 'HalfMegaByte' is not defined

```

```
ls -l /mnt/ltfs/A00236JD/1/2/3/ref05.log
-rwxrwxrwx. 1 root root 0 Jun  5 02:00 /mnt/ltfs/A00236JD/1/2/3/ref05.log

```

**Comment by piste-jp-ibm** [comment]
That is the start block number of a source file. `ltfsee_ordered_copy` sorts the files on source tape by `start block number` order and copy them to disk. 

Thus, LTFS can read specified files only with seeking forward. It shall make performance improvement against normal `cp` command.

**Comment by svein** [comment]
> I have updated the copy message like `"/my/cr a cy¥£/path.log" -> "/some destination/my/cr a cy¥£/path.log"`.

Thank you, very helpful. I know ltfscp also had this, but since ltfs_ordered_copy only copy files. What do you think about removing "Copying" from the output for every file copied? It seems a bit redundant to me.

**Comment by piste-jp-ibm** [comment]
I pushed 2 commits.

1. Fix of `undefined` exception (I can't recreate this on my bench even if I created 0-byte files on a tape, so I didn't test this actually)
2. Update copy message like `"path-foo" -> "path-bar"`

**Comment by svein** [comment]
>     1. Fix of `undefined` exception (I can't recreate this on my bench even if I created 0-byte files on a tape, so I didn't test this actually)

That seemed to have fixed the problem

>     2. Update copy message like `"path-foo" -> "path-bar"`

Thank you.
